app/modules/pwm2.c

```diff
@@ -0,0 +1,553 @@
+{
+  iPin->pulseInterruptCcounter = (sPin->pulseResolutions + 1) * sPin->resolutionInterruptCounterMultiplier;
+}
+
+static uint8_t getDutyAdjustment(const uint32_t duty, const uint32_t pulse)
+{
+  if (duty == 0)
+  {
+    return 0;
+  }
+  else if (duty == pulse)
+  {
+    return 2;
+  }
+  else
+  {
+    return 1;
+  }
+}
+
+static void updatePinOffCounter(pwm2_pin_interrupt_t *iPin, pwm2_pin_setup_t *sPin)
+{
+  iPin->offInterruptCounter = (sPin->duty + getDutyAdjustment(sPin->duty, sPin->pulseResolutions)) * sPin->resolutionInterruptCounterMultiplier;
+}
+
+static void reCalculateCommonToAllPinsData(pwm2_module_data_t *moduleData)
+{
+  moduleData->interruptData.enabledGpioMask = findAllEnabledGpioMask(moduleData);
+  moduleData->setupData.interruptTimerCPUTicks = findCommonCPUTicksDivisor(moduleData);
+  moduleData->setupData.interruptTimerTicks = cpuToTimerTicks(moduleData->setupData.interruptTimerCPUTicks);
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup(moduleData, i))
+    {
+      updatePinResolutionToInterruptsMultiplier(&moduleData->setupData.pin[i], moduleData->setupData.interruptTimerCPUTicks);
+      updatePinPulseToInterruptsCounter(&moduleData->interruptData.pin[i], &moduleData->setupData.pin[i]);
+      updatePinOffCounter(&moduleData->interruptData.pin[i], &moduleData->setupData.pin[i]);
+    }
+  }
+}
+
+//############################
+// interrupt handler related
+
+static inline void computeIsPinOn(pwm2_pin_interrupt_t *pin, uint16_t *maskOn)
+{
+  if (pin->currentInterruptCounter == pin->pulseInterruptCcounter)
+  {
+    pin->currentInterruptCounter = 1;
+  }
+  else
+  {
+    pin->currentInterruptCounter++;
+  }
+  // ets_printf("curr=%u    on=%u\n", pin->currentInterruptCounter, (pin->currentInterruptCounter < pin->offInterruptCounter));
+  if (pin->currentInterruptCounter < pin->offInterruptCounter)
+  {
+    *maskOn |= pin->gpioMask;
+  }
+}
+
+static inline bool isPinSetup2(const pwm2_interrupt_handler_data_t *data, const uint8_t pin)
+{
+  return data->pin[pin].gpioMask > 0;
+}
+
+static inline uint16_t findAllPinOns(pwm2_interrupt_handler_data_t *data)
+{
+  uint16_t maskOn = 0;
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup2(data, i))
+    {
+      computeIsPinOn(&data->pin[i], &maskOn);
+    }
+  }
+  return maskOn;
+}
+
+static inline void setGpioPins(const uint16_t enabledGpioMask, const register uint16_t maskOn)
+{
+  GPIO_REG_WRITE(GPIO_OUT_W1TS_ADDRESS, maskOn);
+  const register uint16_t maskOff = ~maskOn & enabledGpioMask;
+  GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, maskOff);
+}
+
+static void ICACHE_RAM_ATTR timerInterruptHandler(os_param_t arg)
+{
+  pwm2_interrupt_handler_data_t *data = (pwm2_interrupt_handler_data_t *)arg;
+  setGpioPins(data->enabledGpioMask, findAllPinOns(data));
+}
+
+//############################
+// lua bindings
+
+static int lpwm2_open(lua_State *L)
```

Comment:
Please read PiL 28.1 and review userdata use.  This should using `lua_newuserdata()` etc. instead of malloc.  See `app/lua/luac_cross/liolib.c` for example.  The meta method `__GC` allows you to integrate close down and cleanup into the Lua disposal system.

If you are using an OO paradigm then the MD is maintained by the calling application as an inscope variable and the methods called with the `:` syntax.  If you want to use a static singleton then you need to store a reference in the Registry to ensure that it is referenced for GC purposes.

Reply:
i think i'm getting what you're saying. and i thought of same at some point of time but then i opted towards more procedural way.
if you look at the lua api of this module, do you really see a need to introduce client object? 
to me at least current interaction pattern seems consistent enough without it.

Reply:
@fikin Nikolay, Even if you adopt a more procedural way and use the `pwm2.` notation, it is still better to stay in the Lua ecosystem in terms of memory allocation and error handling.  For example, the Lua GC is lazy and an allocation failure will kick off a GC and retry.  If you do a plain malloc, then this will just fail.

Reply:
i guess now that there is driver is split from the lua module, this should be ok.